File: NeetCode/Python/Arrays/Easy/242.ValidAnagram.py
# Given two strings s and t, 
# return true if t is an anagram of s, and false otherwise.

def isAnagram(s, t):
    if len(s) != len(t):
        return False
    
    dict = {}

    for letter in s:
        dict[letter] = dict.get(letter, 0) + 1

    for letter in t:
        if dict.get(letter, 0) == 0:
            return False
        else:
            dict[letter] -= 1

    return True

print(isAnagram("anagram", "nagaram"))
print(isAnagram("rat", "car"))
print(isAnagram("aacc", "ccac"))

File: NeetCode/Swift/Arrays/Easy/242.ValidAnagram.swift
//Given two strings s and t, 
//return true if t is an anagram of s, and false otherwise.

func isAnagram(_ s: String, _ t: String) -> Bool {
    if s.count != t.count {
        return false
    }
    
    var dict = s.reduce(into: [Character: Int]()) { result, letter in
        result[letter, default: 0] += 1
    }
    
    return t.allSatisfy { letter -> Bool in
        if let count = dict[letter], count > 0 {
            dict[letter] = count - 1
            return true
        } else {
            return false
        }
    }
}

print(isAnagram("anagram", "nagaram"))
print(isAnagram("rat", "car"))
print(isAnagram("aacc", "ccac"))